### PR TITLE
Improve performance of WithinStarlaneJumps.

### PR DESCRIPTION
**Pull request by LGM-Doyle**
The WithinStarlaneJumps condition determines if objects are within j jumps of a candidate object.  It currently does that by determining the jump distance between the candidate and each object in the set.  This potentially involves a cache miss for every system associated with an object in the set.

This PR modifies the Pathfinder to allow a single cache hit to return the distances from one system to all others in a single cache hit and a single pathfind.

There are two significant changes.
- It creates `examine_row()` which allows a single `distance_cache_matrix` hit to return the entire row.  A single cache hit will determine the distance from any one system to every other system.
- It abstracts object location into `ObjectSystemIDType` which is a boost::variant encompassing 3 concepts of location: nowhere, one system and in transit.  It then uses variant visitors to determine the distances between these abstracted locations.

The `WithinJumpsOfOthers()` uses the abstracted location, the visitors and the ability to examine an entire row to compute its result quickly.

### CODE REVIEW

File: universe/Pathfinder.cpp [hunk 1]
```diff
@@ -907,6 +1011,148 @@ std::multimap<double, int> Pathfinder::PathfinderImpl::ImmediateNeighbors(int sy
     return std::multimap<double, int>();
 }
 
+
+/** Examine a single universe object and determine if it is within jumps
+    of any object in others.*/
+struct WithinJumpsOfOthersObjectVisitor : public boost::static_visitor<bool> {
+    WithinJumpsOfOthersObjectVisitor(Pathfinder::PathfinderImpl const & _pf,
+                             int _jumps,
```
Comment (geoffthemedio) [inline universe/Pathfinder.cpp]: funky indenting...

File: universe/Pathfinder.cpp [hunk 2]
```diff
@@ -907,6 +1011,148 @@ std::multimap<double, int> Pathfinder::PathfinderImpl::ImmediateNeighbors(int sy
...
+    Pathfinder::PathfinderImpl const & pf;
+    int jumps;
+    std::vector<std::shared_ptr<const UniverseObject> > const & others;
+};
+
+/** Examine a single other in the cache to see if any of its locations
+    are within jumps*/
+struct WithinJumpsOfOthersOtherVisitor : public boost::static_visitor<bool> {
+    WithinJumpsOfOthersOtherVisitor(Pathfinder::PathfinderImpl const & _pf,
+                            int _jumps,
```
Comment (geoffthemedio) [inline universe/Pathfinder.cpp]: and here

File: universe/Pathfinder.cpp [hunk 3]
```diff
@@ -907,6 +1011,148 @@ std::multimap<double, int> Pathfinder::PathfinderImpl::ImmediateNeighbors(int sy
...
+            || single_result(prev_next.second);
+    }
+    Pathfinder::PathfinderImpl const & pf;
+    int jumps;
+    const distance_matrix_storage<short>::row_ref row;
+};
+
+
+void Pathfinder::PathfinderImpl::WithinJumpsOfOthersCacheHit(
+    bool * const answer,
```
Comment (geoffthemedio) [inline universe/Pathfinder.cpp]: remove space b/w bool and *
also prefer const before bool

File: universe/Pathfinder.cpp [hunk 4]
```diff
@@ -907,6 +1011,148 @@ std::multimap<double, int> Pathfinder::PathfinderImpl::ImmediateNeighbors(int sy
...
+    int jumps,
+    std::vector<std::shared_ptr<const UniverseObject> > const & others,
+    size_t ii, const distance_matrix_storage<short>::row_ref row) const {
```
Comment (geoffthemedio) [inline universe/Pathfinder.cpp]: { to newline after multi-line declaration

File: universe/Pathfinder.cpp [hunk 5]
```diff
@@ -907,6 +1011,148 @@ std::multimap<double, int> Pathfinder::PathfinderImpl::ImmediateNeighbors(int sy
...
+    Pathfinder::PathfinderImpl const & pf;
+    int jumps;
+    const distance_matrix_storage<short>::row_ref row;
+};
+
+
+void Pathfinder::PathfinderImpl::WithinJumpsOfOthersCacheHit(
+    bool * const answer,
+    int jumps,
+    std::vector<std::shared_ptr<const UniverseObject> > const & others,
```
Comment (geoffthemedio) [inline universe/Pathfinder.cpp]: > > to >>

File: universe/Pathfinder.cpp [hunk 6]
```diff
@@ -907,6 +1011,148 @@ std::multimap<double, int> Pathfinder::PathfinderImpl::ImmediateNeighbors(int sy
...
+    size_t ii, const distance_matrix_storage<short>::row_ref row) const {
+
+    // Check if any of the others are within jumps of candidate, by looping
+    // through all of the others and applying the WithinJumpsOfOthersOtherVisitor.
+    *answer = false;
+    for (const auto& other : others) {
```
Comment (geoffthemedio) [inline universe/Pathfinder.cpp]: variable names could be improved...

File: universe/Pathfinder.cpp [hunk 7]
```diff
@@ -907,6 +1011,148 @@ std::multimap<double, int> Pathfinder::PathfinderImpl::ImmediateNeighbors(int sy
...
+        }
+    }
+}
+
+void Pathfinder::WithinJumpsOfOthers(
+    int jumps,
+    std::vector<std::shared_ptr<const UniverseObject> > & near,
+    std::vector<std::shared_ptr<const UniverseObject> > & far,
+    std::vector<std::shared_ptr<const UniverseObject> > & candidates,
+    std::vector<std::shared_ptr<const UniverseObject> > const & others) const {
```
Comment (geoffthemedio) [inline universe/Pathfinder.cpp]: { to newline

File: universe/Pathfinder.cpp [hunk 8]
```diff
@@ -907,6 +1011,148 @@ std::multimap<double, int> Pathfinder::PathfinderImpl::ImmediateNeighbors(int sy
...
+    pimpl->WithinJumpsOfOthers(jumps, near, far, candidates, others);
+}
+
+void Pathfinder::PathfinderImpl::WithinJumpsOfOthers(
+    int jumps,
+    std::vector<std::shared_ptr<const UniverseObject> > & near,
```
Comment (geoffthemedio) [inline universe/Pathfinder.cpp]: > >

File: universe/Pathfinder.cpp [hunk 9]
```diff
@@ -907,6 +1011,148 @@ std::multimap<double, int> Pathfinder::PathfinderImpl::ImmediateNeighbors(int sy
...
+    std::vector<std::shared_ptr<const UniverseObject> > & far,
+    std::vector<std::shared_ptr<const UniverseObject> > & candidates_,
+    std::vector<std::shared_ptr<const UniverseObject> > const & others) const {
```
Comment (geoffthemedio) [inline universe/Pathfinder.cpp]: space b4 &
prefer const before type name

File: universe/Pathfinder.cpp [hunk 10]
```diff
@@ -907,6 +1011,148 @@ std::multimap<double, int> Pathfinder::PathfinderImpl::ImmediateNeighbors(int sy
...
+        if (is_near)
+            near.push_back(candidate);
+        else
+            far.push_back(candidate);
+    }
+}
+
+bool Pathfinder::PathfinderImpl::WithinJumpsOfOthers(
+    int jumps, int system_id,
+    std::vector<std::shared_ptr<const UniverseObject> > const & others) const {
```
Comment (geoffthemedio) [inline universe/Pathfinder.cpp]: {

File: universe/Pathfinder.cpp [hunk 11]
```diff
@@ -907,6 +1012,148 @@ std::multimap<double, int> Pathfinder::PathfinderImpl::ImmediateNeighbors(int sy
...
+                                     std::vector<std::shared_ptr<const UniverseObject>> const & _others
+                                    ) :
+        pf(_pf), jumps(_jumps), others(_others) {}
+
+    bool operator()(NowhereType) const { return false; }
+    bool operator()(int sys_id) const {
+        bool retval = pf.WithinJumpsOfOthers(jumps, sys_id, others);
+        return retval;
+    }
+    bool operator()(std::pair<int, int> prev_next) const {
```
Comment (geoffthemedio) [inline universe/Pathfinder.cpp:1029]: does matter between const byref and byvalue for just a pair<int, int> ?

**Review by LGM-Doyle — COMMENTED** [review]
I think that I have made all of the changes.

File: universe/Pathfinder.cpp [hunk 12]
```diff
@@ -907,6 +1012,148 @@ std::multimap<double, int> Pathfinder::PathfinderImpl::ImmediateNeighbors(int sy
...
+};
+
+/** Examine a single other in the cache to see if any of its locations
+    are within jumps*/
+struct WithinJumpsOfOthersOtherVisitor : public boost::static_visitor<bool> {
+    WithinJumpsOfOthersOtherVisitor(Pathfinder::PathfinderImpl const & _pf,
+                                    int _jumps,
+                                    const distance_matrix_storage<short>::row_ref _row
+                                   ) :
+        pf(_pf), jumps(_jumps), row(_row) {}
```
Comment (geoffthemedio) [inline universe/Pathfinder.cpp]: prefer {} on new line

File: universe/Pathfinder.cpp [hunk 13]
```diff
@@ -907,6 +1012,148 @@ std::multimap<double, int> Pathfinder::PathfinderImpl::ImmediateNeighbors(int sy
...
+    try {
+        system_index = m_system_id_to_graph_index.at(system_id);
+    } catch (const std::out_of_range& e) {
+        ErrorLogger() << "Passed invalid system id: " << system_id;
+        return false;
+    }
+
+    // Examine the cache to see if \p system_id is within \p jumps of \p others
+    bool within_jumps(false);
+    distance_matrix_cache< distance_matrix_storage<short>> cache(m_system_jumps);
```
Comment (geoffthemedio) [inline universe/Pathfinder.cpp]: prefer no space before distance

File: universe/Pathfinder.h [hunk 14]
```diff
@@ -101,8 +101,20 @@ class FO_COMMON_API Pathfinder {
...
+    //TODO empire_id is never set to anything other than self, which in
+    //the AI's is the same as ALL_EMPIRES
     std::multimap<double, int>              ImmediateNeighbors(int system_id, int empire_id = ALL_EMPIRES) const;
 
+    /** Returns the \p near and \p far objects from the \p candidates that are
+        within \p jumps of the \p others.
+        The \p candidates are modified and transfered to the \near vector.*/
+    void WithinJumpsOfOthers(
+        int jumps,
+        std::vector<std::shared_ptr<const UniverseObject> > & near,
```
Comment (geoffthemedio) [inline universe/Pathfinder.h]: `> >`

File: universe/Pathfinder.cpp [hunk 15]
```diff
@@ -627,122 +698,156 @@ short Pathfinder::PathfinderImpl::JumpDistanceBetweenSystems(int system1_id, int
...
         }
+
+        int jumps1 = (sjumps1 == -1) ? INT_MAX : static_cast<int>(sjumps1);
+        int jumps2 = (sjumps2 == -1) ? INT_MAX : static_cast<int>(sjumps2);
+
+        int jumps = std::min(jumps1, jumps2);
+        return jumps;
     }
-    return INT_MAX;
+    Pathfinder::PathfinderImpl const & pf;
```
Comment (geoffthemedio) [inline universe/Pathfinder.cpp]: prefer `const type& var;`

**Comment by geoffthemedio** [comment]
where do you invalidate cached results if the positions or connectivity of systems changes? or once per turn regardless?

**Comment by LGM-Doyle** [comment]
The function `Pathfinder::InitializeSystemGraph()` invalidates the cache.  

`InitializeSystemGraph()` builds a graph of the galaxy starlanes and compares it with the current graph.  If it is different then it clears the cache in line https://github.com/freeorion/freeorion/blob/master/universe/Pathfinder.cpp#L1008.

`InitializeSystemGraph()` is called in a number of places:
- in universe generation after starlanes are added 😲 : https://github.com/freeorion/freeorion/blob/master/universe/UniverseGenerator.cpp#L830
- AI on game start https://github.com/freeorion/freeorion/blob/master/client/AI/AIClientApp.cpp#L251
- AI on turn update https://github.com/freeorion/freeorion/blob/master/client/AI/AIClientApp.cpp#L336
- Server on load game https://github.com/freeorion/freeorion/blob/master/server/ServerApp.cpp#L1129
- Server in post combat processing https://github.com/freeorion/freeorion/blob/master/server/ServerApp.cpp#L2838
- UI in IniTurn https://github.com/freeorion/freeorion/blob/master/UI/MapWnd.cpp#L2311
- UI in MidTurn update https://github.com/freeorion/freeorion/blob/master/UI/MapWnd.cpp#L2473

Until Starlane Bore tech is used or the Experimenters activate, the cache may not be invalidated.

**Comment by geoffthemedio** [comment]
What if systems move but their lane connectivity doesn't change? Would that be detected and accounted for in choosing the shortest distance path?

**Comment by LGM-Doyle** [comment]
No it did not.  The comparison in the legacy code was for topology only.

I've added a comparison for edge weight.  If the length of a starlane changes, then the whole graph is invalidated.

File: universe/Pathfinder.h [hunk 16]
```diff
@@ -101,8 +101,20 @@ class FO_COMMON_API Pathfinder {
       * \a empire_id == ALL_EMPIRES.
       * \throw std::out_of_range This function will throw if the  system
       * ID is out of range. */
+    //TODO empire_id is never set to anything other than self, which in
+    //the AI's is the same as ALL_EMPIRES
     std::multimap<double, int>              ImmediateNeighbors(int system_id, int empire_id = ALL_EMPIRES) const;
 
+    /** Returns the \p near and \p far objects from the \p candidates that are
+        within \p jumps of the \p others.
+        The \p candidates are modified and transfered to the \near vector.*/
```
Comment (geoffthemedio) [inline universe/Pathfinder.h]: confused... do objects also get moved into `far`? is anything left in `candidates`?

don't like `others` as a parameter name... too vague. perhaps `from_objects` ?

File: universe/Pathfinder.h [hunk 17]
```diff
@@ -124,7 +136,6 @@ class FO_COMMON_API Pathfinder {
 
     //@}
 
-private:
```
Comment (geoffthemedio) [inline universe/Pathfinder.h:127]: why not private?

File: universe/Pathfinder.cpp [hunk 18]
```diff
@@ -907,6 +1012,149 @@ std::multimap<double, int> Pathfinder::PathfinderImpl::ImmediateNeighbors(int sy
...
+            || single_result(prev_next.second);
+    }
+    const Pathfinder::PathfinderImpl& pf;
+    int jumps;
+    const distance_matrix_storage<short>::row_ref row;
+};
+
+
+void Pathfinder::PathfinderImpl::WithinJumpsOfOthersCacheHit(
+    bool& answer,
```
Comment (geoffthemedio) [inline universe/Pathfinder.cpp]: why by reference and not as return value?

File: universe/Pathfinder.cpp [hunk 19]
```diff
@@ -907,6 +1012,149 @@ std::multimap<double, int> Pathfinder::PathfinderImpl::ImmediateNeighbors(int sy
...
+    std::vector<std::shared_ptr<const UniverseObject>> const & others;
+};
+
+/** Examine a single other in the cache to see if any of its locations
+    are within jumps*/
+struct WithinJumpsOfOthersOtherVisitor : public boost::static_visitor<bool> {
+    WithinJumpsOfOthersOtherVisitor(const Pathfinder::PathfinderImpl& _pf,
+                                    int _jumps,
+                                    const distance_matrix_storage<short>::row_ref _row
+                                   ) :
```
Comment (geoffthemedio) [inline universe/Pathfinder.cpp]: probably put at end of preceeding line

File: universe/Pathfinder.cpp [hunk 20]
```diff
@@ -907,6 +1012,149 @@ std::multimap<double, int> Pathfinder::PathfinderImpl::ImmediateNeighbors(int sy
...
+    bool operator()(NowhereType) const { return false; }
+    bool operator()(int sys_id) const {
+        bool retval = pf.WithinJumpsOfOthers(jumps, sys_id, others);
+        return retval;
+    }
+    bool operator()(std::pair<int, int> prev_next) const {
+        return pf.WithinJumpsOfOthers(jumps, prev_next.first, others)
+            || pf.WithinJumpsOfOthers(jumps, prev_next.second, others);
+    }
+    Pathfinder::PathfinderImpl const & pf;
```
Comment (geoffthemedio) [inline universe/Pathfinder.cpp]: `const type& varname` preferred

File: universe/Pathfinder.cpp [hunk 21]
```diff
@@ -907,6 +1012,149 @@ std::multimap<double, int> Pathfinder::PathfinderImpl::ImmediateNeighbors(int sy
...
 
+
+/** Examine a single universe object and determine if it is within jumps
+    of any object in others.*/
+struct WithinJumpsOfOthersObjectVisitor : public boost::static_visitor<bool> {
+    WithinJumpsOfOthersObjectVisitor(Pathfinder::PathfinderImpl const & _pf,
+                                     int _jumps,
+                                     std::vector<std::shared_ptr<const UniverseObject>> const & _others
+                                    ) :
+        pf(_pf), jumps(_jumps), others(_others) {}
```
Comment (geoffthemedio) [inline universe/Pathfinder.cpp]: `{}` to newline
`) :` to previous line
`const type& varname` preferred

File: universe/Pathfinder.cpp [hunk 22]
```diff
@@ -627,122 +698,156 @@ short Pathfinder::PathfinderImpl::JumpDistanceBetweenSystems(int system1_id, int
...
-                // no path
-                return INT_MAX;
-            }
+/** JumpDistanceSys1Visitor visits the first system and uses
+    JumpDistanceSysVisitor2 to determines the distance between \p _sys_id2 and
+    the GeneralizedLocation that it is visiting.*/
+struct JumpDistanceSys1Visitor : public boost::static_visitor<int> {
+    JumpDistanceSys1Visitor(Pathfinder::PathfinderImpl const & _pf,
+                            GeneralizedLocationType const &_sys2_ids) :
+        pf(_pf), sys2_ids(_sys2_ids) {}
```
Comment (geoffthemedio) [inline universe/Pathfinder.cpp]: `{}` to newline

File: universe/Pathfinder.cpp [hunk 23]
```diff
@@ -627,122 +698,156 @@ short Pathfinder::PathfinderImpl::JumpDistanceBetweenSystems(int system1_id, int
...
-                return jumps - 1;
-            } else {
-                // no path
-                return INT_MAX;
-            }
-        }
+        int jumps = std::min(jumps1, jumps2);
+        return jumps;
+   }
+    Pathfinder::PathfinderImpl const & pf;
```
Comment (geoffthemedio) [inline universe/Pathfinder.cpp]: const type& varname

(I probably won't note all such cases)

File: universe/Pathfinder.cpp [hunk 24]
```diff
@@ -627,122 +698,156 @@ short Pathfinder::PathfinderImpl::JumpDistanceBetweenSystems(int system1_id, int
...
+       Variant in boost::optional allows the visitors to be more
+       uniform.  The type needed to be not implicitly convertible to
+       int.  The type should really be void.
+
+    */
+
+    // A constant to indate a UniverseObject is nowhere.
+    typedef void (*NowhereType)();
+    void Nowhere() {}
+    typedef boost::variant<NowhereType, int, std::pair<int, int >> GeneralizedLocationType;
```
Comment (geoffthemedio) [inline universe/Pathfinder.cpp]: extra space between `int >>`

File: universe/Pathfinder.cpp [hunk 25]
```diff
@@ -627,122 +698,156 @@ short Pathfinder::PathfinderImpl::JumpDistanceBetweenSystems(int system1_id, int
...
+
+    /** Return the location of \p obj.*/
+    GeneralizedLocationType GeneralizedLocation(std::shared_ptr<const UniverseObject> obj) {
+        if (!obj)
+            return &Nowhere;
```
Comment (geoffthemedio) [inline universe/Pathfinder.cpp]: returning the address of a function...? seems odd...

File: universe/Condition.cpp [hunk 26]
```diff
@@ -6680,9 +6652,18 @@ bool WithinStarlaneJumps::Match(const ScriptingContext& local_context) const {
...
+
     int jump_limit = m_jumps->Eval(local_context);
+    if (jump_limit < 0)
+        return false;
 
-    return WithinStarlaneJumpsSimpleMatch(subcondition_matches, jump_limit)(candidate);
+    ObjectSet near;
+    ObjectSet one;
+    one.push_back(candidate);
+    GetUniverse().GetPathfinder()->WithinJumpsOfOthers(jump_limit, near, one, one, subcondition_matches);
```
Comment (geoffthemedio) [inline universe/Condition.cpp]: explanitory comment for parameters required...

File: universe/Condition.cpp [hunk 27]
```diff
@@ -6636,8 +6606,10 @@ void WithinStarlaneJumps::Eval(const ScriptingContext& parent_context,
         ObjectSet subcondition_matches;
         m_condition->Eval(local_context, subcondition_matches);
         int jump_limit = m_jumps->Eval(local_context);
+        ObjectSet &from_set( search_domain == Condition::MATCHES ? matches : non_matches);
```
Comment (geoffthemedio) [inline universe/Condition.cpp]: space after ( seems odd

**Comment by geoffthemedio** [comment]
Do you have any anecdotal measurements of speed improvement from this?

Builds and seems to work OK for me.

**Review by LGM-Doyle — COMMENTED** [review]
On my standard test case 999 systems with 39 AIs this PR  provides an improvement of 3.5 seconds on a 42 second startup time.  

The method of using `examine_row` to vectorize other operations, will provide other small improvements.

File: universe/Condition.cpp [hunk 28]
```diff
@@ -6680,9 +6652,22 @@ bool WithinStarlaneJumps::Match(const ScriptingContext& local_context) const {
...
     m_condition->Eval(local_context, subcondition_matches);
+    if (subcondition_matches.empty())
+        return false;
+
     int jump_limit = m_jumps->Eval(local_context);
+    if (jump_limit < 0)
+        return false;
+
+    ObjectSet candidate_set;
+    candidate_set.push_back(candidate);
```
Comment (geoffthemedio) [inline universe/Condition.cpp]: `ObjectSet candidate_set{candidate};` could replace the push_back after init

File: universe/Pathfinder.cpp [hunk 29]
```diff
@@ -627,122 +704,150 @@ short Pathfinder::PathfinderImpl::JumpDistanceBetweenSystems(int system1_id, int
...
-                ErrorLogger() << "JumpsBetweenObjects caught exception when calling JumpDistanceBetweenSystems";
-            }
-            int jumps = static_cast<int>(std::max(jumps1, std::max(jumps2, std::max(jumps3, jumps4))));
-            if (jumps != -1) {
-                return jumps - 1;
-            } else {
-                // no path
-                return INT_MAX;
-            }
+        if (next_sys_id!= INVALID_OBJECT_ID) {
```
Comment (geoffthemedio) [inline universe/Pathfinder.cpp:826]: nitpick: space before !

File: universe/Pathfinder.cpp [hunk 30]
```diff
@@ -627,122 +704,150 @@ short Pathfinder::PathfinderImpl::JumpDistanceBetweenSystems(int system1_id, int
...
+        int jumps = std::min(jumps1, jumps2);
+        return jumps;
     }
-    return INT_MAX;
+    const Pathfinder::PathfinderImpl& pf;
+    const GeneralizedLocationType& sys2_ids;
+};
+
+int Pathfinder::JumpDistanceBetweenObjects(int object1_id, int object2_id) const {
+    return pimpl->JumpDistanceBetweenObjects(object1_id, object2_id);
```
Comment (geoffthemedio) [inline universe/Pathfinder.cpp]: nitpick: could put {} on same line as return statement

File: universe/Pathfinder.cpp [hunk 31]
```diff
@@ -907,6 +1012,147 @@ std::multimap<double, int> Pathfinder::PathfinderImpl::ImmediateNeighbors(int sy
...
+    int jumps;
+    const distance_matrix_storage<short>::row_ref row;
+};
+
+
+void Pathfinder::PathfinderImpl::WithinJumpsOfOthersCacheHit(
+    bool& answer,
+    int jumps,
+    const std::vector<std::shared_ptr<const UniverseObject>>& others,
+    size_t ii, const distance_matrix_storage<short>::row_ref row) const
```
Comment (geoffthemedio) [inline universe/Pathfinder.cpp]: having just the last parameter on the same line as the one before it is a bit hard to read... sticking to one-per-line or several-per-line-when-there's-space would be preferable

File: universe/Pathfinder.cpp [hunk 32]
```diff
@@ -907,6 +1012,147 @@ std::multimap<double, int> Pathfinder::PathfinderImpl::ImmediateNeighbors(int sy
...
+}
+
+std::pair<std::vector<std::shared_ptr<const UniverseObject>>, std::vector<std::shared_ptr<const UniverseObject>>>
+Pathfinder::PathfinderImpl::WithinJumpsOfOthers(
+    int jumps,
+    const std::vector<std::shared_ptr<const UniverseObject>>& candidates,
+    const std::vector<std::shared_ptr<const UniverseObject>>& stationary) const
+{
+    // Examine each candidate and transfer those within jumps of the
+    // others into near.
```
Comment (geoffthemedio) [inline universe/Pathfinder.cpp]: ... and those that are not within jumps into far?

File: universe/Pathfinder.cpp [hunk 33]
```diff
@@ -907,6 +1012,147 @@ std::multimap<double, int> Pathfinder::PathfinderImpl::ImmediateNeighbors(int sy
...
+        GeneralizedLocationType candidate_systems = GeneralizedLocation(candidate);
+        bool is_near = boost::apply_visitor(visitor, candidate_systems);
+
+        if (is_near)
+            near.push_back(candidate);
+        else
+            far.push_back(candidate);
+    }
+
+    return std::make_pair(near, far);
```
Comment (geoffthemedio) [inline universe/Pathfinder.cpp]: should be able to do just `{near, far}`

**Comment by geoffthemedio** [comment]
thanks for patience with so many rounds of comments...

File: universe/Pathfinder.cpp (hunk 31 above)
Comment (LGM-Doyle) [reply universe/Pathfinder.cpp]: I changed it to
````
void Pathfinder::PathfinderImpl::WithinJumpsOfOthersCacheHit(
    bool& answer, int jumps,
    const std::vector<std::shared_ptr<const UniverseObject>>& others,
    size_t ii, const distance_matrix_storage<short>::row_ref row) const
````
I was trying to keep that last two parameters on the same line.  Those are the two parameters required for the cache hit handler signature.  All the other parameters need to be bound to something before the cache hit handler is passed to the `distance_matrix_cache`.

**Pull request merged**